packages/devtools_app/lib/src/screens/performance/performance_model.dart

```diff
@@ -1238,77 +1243,112 @@ class FrameAnalysis {
-          .fold<Duration>(Duration.zero, (previous, SyncTimelineEvent event) {
-        return previous + event.time.duration;
-      });
+  void selectFramePhase(FramePhase block) {
+    _selectedPhase.value = block;
+  }
 
-  Duration? _buildTime;
+  /// Data for the build phase of [frame].
+  ///
+  /// This is drawn from all the "Build" events on the UI thread. For a single
+  /// flutter frame, there can be more than one build event, and this data may
+  /// overlap with a portion of the [layoutPhase] if "Build" timeline events are
+  /// children of the "Layout" event.
+  ///
+  /// Example:
+  /// [-----Build----][-----------------Layout-----------------]
+  ///                       [--Build--]     [----Build----]
+  late FramePhase buildPhase = _generateBuildPhase();
+
+  FramePhase _generateBuildPhase() {
+    final uiEvent = frame.timelineEventData.uiEvent;
+    if (uiEvent == null) {
+      return FramePhase.build(events: <SyncTimelineEvent>[]);
+    }
+    final buildEvents = uiEvent
+        .childrenWithCondition(
+          (event) => event.name!.caseInsensitiveEquals(buildEventName),
+        )
+        .cast<SyncTimelineEvent>();
+    return FramePhase.build(events: buildEvents);
+  }
 
-  SyncTimelineEvent? get layoutEvent {
-    if (frame.timelineEventData.uiEvent == null) return null;
-    if (_layoutEvent != null) return _layoutEvent;
+  /// Data for the layout phase of [frame].
+  ///
+  /// This is drawn from the "Layout" timeline event on the UI thread. This data
+  /// may overlap with a portion of the [buildPhase] if "Build" timeline events
+  /// are children of the "Layout" event.
+  ///
+  /// Example:
+  /// [-----------------Layout-----------------]
+  ///    [--Build--]     [----Build----]
+  late FramePhase layoutPhase = _generateLayoutPhase();
 
-    return _layoutEvent =
+  FramePhase _generateLayoutPhase() {
+    if (frame.timelineEventData.uiEvent == null) {
+      return FramePhase.layout(events: <SyncTimelineEvent>[]);
+    }
+    final layoutEvent =
         frame.timelineEventData.uiEvent!.firstChildWithCondition(
       (event) => event.name!.caseInsensitiveEquals(layoutEventName),
-    ) as SyncTimelineEvent?;
+    );
+    return FramePhase.layout(
+      events: <SyncTimelineEvent>[
+        if (layoutEvent != null) layoutEvent as SyncTimelineEvent,
+      ],
+    );
   }
 
-  SyncTimelineEvent? _layoutEvent;
-
-  Duration get layoutTime => layoutEvent?.time.duration ?? Duration.zero;
-
-  SyncTimelineEvent? get paintEvent {
-    if (frame.timelineEventData.uiEvent == null) return null;
-    if (_paintEvent != null) return _paintEvent;
+  /// Data for the Paint phase of [frame].
+  ///
+  /// This is drawn from the "Paint" timeline event on the UI thread
+  late FramePhase paintPhase = _generatePaintPhase();
 
-    return _paintEvent =
-        frame.timelineEventData.uiEvent!.firstChildWithCondition(
+  FramePhase _generatePaintPhase() {
+    if (frame.timelineEventData.uiEvent == null) {
```

Comment:
Maybe pull out into a variable so the null assert isn't need

Reply:
done